blockchain/blockchain.go

```diff
@@ -756,29 +762,43 @@ func (bc *blockchain) startExistingBlockchain() error {
 	return nil
 }
 
-func (bc *blockchain) validateActionsCtx(height uint64) protocol.ValidateActionsCtx {
-	prevBlkHash := bc.tipHash
-	if height == 1 {
-		prevBlkHash = bc.config.Genesis.Hash()
+func (bc *blockchain) tipInfo() (*protocol.TipInfo, error) {
+	if bc.tipHeight == 0 {
+		return &protocol.TipInfo{
+			Height:    0,
+			Hash:      bc.config.Genesis.Hash(),
+			Timestamp: time.Unix(bc.config.Genesis.Timestamp, 0),
+		}, nil
 	}
-	return protocol.ValidateActionsCtx{
-		Genesis: bc.config.Genesis,
-		Tip: protocol.TipInfo{
-			Height:    bc.tipHeight,
-			Hash:      prevBlkHash,
-			Timestamp: time.Now(), // bc.dao.GetBlockTime(tipHeight),
-		},
+	header, err := bc.dao.Header(bc.tipHash)
+	if err != nil {
+		return nil, err
 	}
+
+	return &protocol.TipInfo{
+		Height:    bc.tipHeight,
+		Hash:      bc.tipHash,
+		Timestamp: header.Timestamp(),
+	}, nil
+}
+
+func (bc *blockchain) validateActionsCtx() (*protocol.ValidateActionsCtx, error) {
+	tip, err := bc.tipInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocol.ValidateActionsCtx{Genesis: bc.config.Genesis, Tip: *tip}, nil
 }
 
 func (bc *blockchain) validateBlock(blk *block.Block) error {
 	validateTimer := bc.timerFactory.NewTimer("validate")
-	ctx := protocol.WithValidateActionsCtx(
-		context.Background(),
-		bc.validateActionsCtx(blk.Height()),
-	)
-
-	err := bc.validator.Validate(ctx, blk)
+	vaCtx, err := bc.validateActionsCtx()
+	if err != nil {
+		return err
+	}
+	ctx := protocol.WithValidateActionsCtx(context.Background(), *vaCtx)
```

Comment:
i think can return protocol.ValidateActionsCtx like before, so no need to do *vaCtx here